versions.sh

```diff
@@ -86,12 +86,39 @@ for version in "${versions[@]}"; do
 		jq <<<"$json" -c --argjson doc "$doc" '
 			.[env.version] = {
 				version: env.fullVersion,
-				variants: [ "apache-buster", "fpm-buster", "fpm-alpine3.12" ],
+				variants: [
+					# arrays of arrays to control ordering
+					if [ "7", "8.9", "9.1" ] | index(env.version) then
+						[
+							[ "buster" ],
+							[ "alpine3.13" ]
+						]
+					elif env.version | startswith("9.") then
+						[
+							[ "buster" ],
+							[ "alpine3.14", "alpine3.13" ]
+						]
+					else
+						[
+							[ "bullseye", "buster" ],
+							[ "alpine3.14", "alpine3.13" ]
+						]
+					end
+					| . as $os
+					| [
+						[ "apache", "fpm" ],
+						[ "fpm" ]
+					] as $variant
+					| keys[]
+					| $variant[.][] as $var | $os[.][] | [ $var, . ] | join("-")
```

Comment:
Is this section overkill to generate the variants?

```jq
# 9.2
    "variants": [
      "apache-buster",
      "fpm-buster",
      "fpm-alpine3.14",
      "fpm-alpine3.13"
    ],
# for 10+
    "variants": [
      "apache-bullseye",
      "fpm-bullseye",
      "apache-buster",
      "fpm-buster",
      "fpm-alpine3.14",
      "fpm-alpine3.13"
    ],
```

Reply:
It's definitely complicated and a bit hard to follow -- IMO this would be a little more obvious if we flatten the lists above, and then only add the `apache` prefix specially, something like:

```jq
if [ "7", "8.9", "9.1" ] | index(env.version) then
	[
		"buster",
		"alpine3.13"
	]
elif env.version | startswith("9.") then
	[
		"buster",
		#"alpine3.14", "alpine3.13"
		# sometimes I go overkill here with something like:
		(
			"3.14",
			"3.13"
		| "alpine" + .)
	]
else
	[
		"bullseye", "buster",
		"alpine3.14", "alpine3.13"
	]
end
| map(
	if startswith("alpine") then empty else "apache-" + . end,
	"fpm-" + .
)
```